api/api.go

```diff
@@ -0,0 +1,93 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/consul-terraform-sync/event"
+)
+
+const (
+	defaultAPIVersion = "v1"
+
+	statusHealthy      = "healthy"
+	statusDegraded     = "degraded"
+	statusCritical     = "critical"
+	statusUndetermined = "undetermined"
+)
+
+// API supports api requests to the cts biniary
+type API struct {
+	store   *event.Store
+	port    int
+	version string
+	srv     *http.Server
+}
+
+// NewAPI create a new API object
+func NewAPI(store *event.Store, port int) *API {
+	mux := http.NewServeMux()
+
+	// retrieve task status for a task-name
+	mux.Handle(fmt.Sprintf("/%s/%s/", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(store, defaultAPIVersion))
+	// retrieve all task statuses
+	mux.Handle(fmt.Sprintf("/%s/%s", defaultAPIVersion, taskStatusPath),
+		newTaskStatusHandler(store, defaultAPIVersion))
+
+	srv := &http.Server{
+		Addr:         fmt.Sprintf(":%d", port),
+		WriteTimeout: time.Second * 15,
+		ReadTimeout:  time.Second * 15,
+		IdleTimeout:  time.Second * 60,
+		Handler:      mux,
+	}
+
+	return &API{
+		port:    port,
+		store:   store,
+		version: defaultAPIVersion,
+		srv:     srv,
+	}
+}
+
+// Serve starts up and handles shutdown for the http server to serve
+// API requests
+func (api *API) Serve(ctx context.Context) {
+	go func() {
+		if err := api.srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			// go ahead and stop cts from continuing
+			log.Fatalf("error starting api server at '%d': '%s'\n", api.port, err)
+		}
+	}()
+
+	log.Printf("[INFO] (api) server started at port %d", api.port)
+
+	ctxShutDown, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Printf("[INFO] (api) stopping api server")
+				if err := api.srv.Shutdown(ctxShutDown); err != nil {
+					log.Printf("[ERROR] (api) error stopping api server: '%s'", err)
+				}
+				cancel()
+				return
+			}
+		}
+	}()
+}
+
+// jsonResponse adds the return response for handlers
+func jsonResponse(w http.ResponseWriter, code int, response interface{}) {
+	w.Header().Add("Content-Type", "application/json")
+	w.WriteHeader(code)
+	json.NewEncoder(w).Encode(response)
```

Comment:
As all responses are supplied internally and can be sure to be encode-able w/o error, I don't see a problem with ignoring the error here as long as all the calls using this have their json encoding tested.

Reply:
Thanks for pointing this out! I went with writing a test to test the json encoding. It ended up being more code overall but kept the handler code tidier 🤔.